Pass base rest config into client factory

nrfox opened the pull request:
### Describe the change

Refactors the client factory to accept a `rest.Config` instead of calling `InClusterConfig()` inside of it. This makes it much easier to use an alternative to the in cluster config when running outside of the cluster such as with `run-kiali.sh` or local mode. Also removes global vars from the client factory.

### Steps to test the PR

1. Regression test.
2. Run the `run-kiali.sh` hack script to ensure it still works with this change.

### Automation testing

N/A

### Issue reference

Part of #8274

nrfox commented:
> Just looking at the code, there are a bunch of changes reworking the recent logic applied for external Kiali support.

The biggest change I think is this one: https://github.com/kiali/kiali/pull/8611/files#diff-e23b07551e997576f48ef016d1d58ffe1d71854c5846c12a1cac69995621bcd1L102-L109. Now since we're passing in the rest.Config we don't need to assume that on error we're running outside the cluster. You get that for free when you call `ctrl.GetConfig()`: https://github.com/kiali/kiali/pull/8611/files#diff-a0c270f708064531f39882e8d88e4fbf3d0a2e9f9adf7a034903ed036518ac79R135-R138

Internally `ctrl.GetConfig` looks for a kubeconfig flag then looks for an env var named `KUBECONFIG`, which we now set in `run-kiali`, and if it doesn't find those then it tries the InClusterConfig.

`run-kiali` and local mode are basically the same thing except all the logic is baked into the Kiali server with local mode.

> It's not totally clear to me how run-kiali will work to allow a user to run the server in the IDE, but I'll give it all a try.

I run the script to setup all the port forwarding without starting the server and run a debug session of the server which can use all the services that run-kiali setups. With local mode you just execute kiali with `kiali run` and it does the rest.

jshaughn reviewed (commented):
I revoked my initial comment, I did a totally clean build and I think it may have been an env issue. Continuing with testing...

... run-kiali in an external kiali env does not seem to be behaving correctly. It does start, but the UI is showing unusual results. With the current context set to `mesh`, I start run-kiali like this:

`$ hack/run-kiali.sh --enable-server false --tmp-root-dir $HOME/tmp -pg 3005:3005 -kc mesh -cn mgmt -hkc mgmt`

Below, you can see bookinfo namespace on both clusters (it's only on `mesh`), and also how it thinks there is a control plane on each cluster (again, only on `mesh`):
<img width="1711" height="695" alt="image" src="https://github.com/user-attachments/assets/14bde8e8-54e1-46f8-abf0-23725cf0ec9b" />

Master looks correct:
<img width="1711" height="695" alt="image" src="https://github.com/user-attachments/assets/ab19db37-718e-4897-9739-5662bb2c0e20" />

This was probably the trickiest use case in the existing code for external kiali. Not sure if I need different args or we need to tweak something in the logic. Let me know if there is something you want me to try.

nrfox commented:
I think there's something weird with how `run-kiali` is being setup for external deployment. This is from local mode which is rebased on top of this change:
<img width="2351" height="1309" alt="Screenshot from 2025-08-05 19-32-20" src="https://github.com/user-attachments/assets/15cde526-5f79-4399-a1c2-bedc9da6aaa3" />

which does not include `bookinfo` from both clusters.

I'll take a look at that script and see what's going on.

nrfox commented:
@jshaughn I added one arg to the run-kiali script and ran it like this against the environment setup locally by your CI PR:
```
hack/run-kiali.sh --kiali-exe ~/go/bin/kiali --kube-context current --copy-cluster-secrets true --istiod-url "" --ignore-home-cluster true --cluster-name mgmt --enable-server false
```
`--istiod-url ""` just tells Kiali not to port-forward to istiod in the `mgmt` cluster because there is no istiod in the mgmt cluster.

jshaughn commented:
>  I added one arg to the run-kiali script and ran it like this against the environment setup locally by your CI PR:

I'll give it a try!

nrfox commented:
Ok after rebasing with master running:
```
sudo rm -rf /kiali-remote-cluster-secrets/
hack/run-integration-tests.sh --test-suite frontend-external-kiali --setup-only true
```

I then ran:
```
hack/run-kiali.sh --kiali-exe ~/go/bin/kiali --kube-context current --copy-cluster-secrets true --istiod-url "" --ignore-home-cluster true
```
Which fails because there's no prometheus on the mgmt cluster. Something we we're going to address in a followup to #8608.
```
📄 You are connecting to a (non-OpenShift) Kubernetes cluster
Error from server (NotFound): deployments.apps "prometheus" not found
🚨 ERROR: Cannot auto-discover Prometheus on Kubernetes. You must specify the Prometheus URL via --prometheus-url
```

If I deploy all the external services to the mgmt cluster:
```
kubectl --context kind-mgmt apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.26/samples/addons/prometheus.yaml -n istio-system
kubectl --context kind-mgmt apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.26/samples/addons/jaeger.yaml -n istio-system
kubectl --context kind-mgmt apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.26/samples/addons/grafana.yaml -n istio-system
```
And then run:
```
hack/run-kiali.sh --kiali-exe ~/go/bin/kiali --kube-context current --copy-cluster-secrets true --istiod-url "" --ignore-home-cluster true --cluster-name mgmt
```
It works for me:
<img width="1735" height="1005" alt="Screenshot from 2025-08-06 11-56-00" src="https://github.com/user-attachments/assets/1574b25a-4f3d-4e61-a2d0-e5d705b4f653" />
<img width="2325" height="1114" alt="Screenshot from 2025-08-06 12-02-32" src="https://github.com/user-attachments/assets/8929cb6d-d857-4e60-8c57-183c629c8813" />

Here's what my remote secrets dir looks like:
```
ls /kiali-remote-cluster-secrets/
kiali-multi-cluster-secret/  kiali-remote-cluster-secret-mesh/  mgmt/
```

nrfox commented:
I have added the addons to the mgmt cluster and updated Kiali to use these. I think there's still some problems with how the addons are configured, like traces not being exported from mesh --> mgmt but at least you can run the hack script without errors now.

These are the steps to test:
```
sudo rm -rf /kiali-remote-cluster-secrets/
hack/run-integration-tests.sh --test-suite frontend-external-kiali --setup-only true
hack/run-kiali.sh --kiali-exe ~/go/bin/kiali --kube-context current --copy-cluster-secrets true --istiod-url "" --ignore-home-cluster true
```

And this is what you should see:
<img width="2327" height="1102" alt="Screenshot from 2025-08-07 12-19-41" src="https://github.com/user-attachments/assets/ba0db818-ca77-4ff6-bcda-5ce53f44f654" />
<img width="2324" height="1153" alt="Screenshot from 2025-08-07 12-19-30" src="https://github.com/user-attachments/assets/0afa7a4b-10b0-472c-a6dd-172d39c60be8" />

jshaughn reviewed (commented):
Everything is now looking good with setup with minikube, using the `install-external-kiali.sh` directly like:

`$ hack/istio/multicluster/install-external-kiali.sh -kudi true -kshc ~/go/src/github.com/kiali/kiali/helm-charts/_output/charts/kiali-server-2.14.0-SNAPSHOT.tgz`
```

$ kubectl --context mgmt -n istio-system get pods
NAME                          READY   STATUS    RESTARTS   AGE
grafana-65bfb5f855-22w4l      1/1     Running   0          36m
jaeger-868fbc75d7-k7spz       1/1     Running   0          36m
kiali-775d8c457b-95wdm        1/1     Running   0          36m
prometheus-689cc795d4-4949z   2/2     Running   0          36m
~/go/src/github.com/kiali/kiali$ kubectl --context mgmt -n bookinfo get pods
No resources found in bookinfo namespace.
~/go/src/github.com/kiali/kiali$ kubectl --context mesh -n istio-system get pods
NAME                                    READY   STATUS    RESTARTS   AGE
istio-ingressgateway-86cdf57597-2k4xr   1/1     Running   0          37m
istiod-84c4b467d4-8c9hh                 1/1     Running   0          37m
jaeger-868fbc75d7-xsdsk                 1/1     Running   0          37m
prometheus-689cc795d4-qnmd8             2/2     Running   0          37m
~/go/src/github.com/kiali/kiali$ kubectl --context mesh -n bookinfo get pods
NAME                              READY   STATUS    RESTARTS        AGE
details-v1-766844796b-x7slt       2/2     Running   0               37m
kiali-traffic-generator-hxhrq     1/1     Running   0               2m36s
mongodb-v1-7b54f4bcc9-62rjt       2/2     Running   1 (2m40s ago)   3m
productpage-v1-54bb874995-jrj2g   2/2     Running   0               37m
ratings-v1-5dc79b6bcd-vnm9c       2/2     Running   0               37m
ratings-v2-5484bc5f99-l2qt7       2/2     Running   0               3m
reviews-v1-598b896c9d-q5mn8       2/2     Running   0               37m
reviews-v2-556d6457d-rnw4d        2/2     Running   0               37m
reviews-v3-564544b4d6-lz5np       2/2     Running   0               37m
```
Using the UI directly (not run-kiali) looks good:
<img width="1701" height="523" alt="image" src="https://github.com/user-attachments/assets/2b6dd76f-e95e-4a6f-9064-128dbb0558c6" />
<img width="1325" height="617" alt="image" src="https://github.com/user-attachments/assets/a06a1d49-38bf-4456-8cbc-66410ef5dc01" />
<img width="1075" height="714" alt="image" src="https://github.com/user-attachments/assets/7727ac9a-cfa4-4bb8-b376-31ed25e832de" />

After setting my current context to `mgmt` (it was set to `mesh` after running the script), I was able to use the suggested command, but I added `--enable-server false` in order to test running the server in debug:

`hack/run-kiali.sh --kiali-exe ~/go/bin/kiali --kube-context current --copy-cluster-secrets true --istiod-url "" --ignore-home-cluster true --cluster-name mgmt --enable-server false`

I'm happy to report that I was able to start a debug server in Cursor and the UI looked good, like the screenshots above.

I'll sanity check the standard co-located install later/tomorrow.

jshaughn commented on hack/istio/multicluster/install-external-kiali.sh:
```diff
@@ -254,21 +235,11 @@ if [ "${KIALI_ENABLED}" == "true" ]; then
+  switch_cluster "${CLUSTER2_CONTEXT}" "${CLUSTER2_USER}" "${CLUSTER2_PASS}"
   echo "Installing bookinfo demo in namespace [${BOOKINFO_NAMESPACE}] on [${CLUSTER2_CONTEXT}]"
-  source ${SCRIPT_DIR}/../install-bookinfo-demo.sh --client-exe "${CLIENT_EXE}" --istio-dir "${ISTIO_DIR}" --istio-namespace "${ISTIO_NAMESPACE}" --namespace "${BOOKINFO_NAMESPACE}" --minikube-profile "${CLUSTER2_CONTEXT}" --namespace "${BOOKINFO_NAMESPACE}" --kube-context "${CLUSTER2_CONTEXT}"
+  source ${SCRIPT_DIR}/../install-bookinfo-demo.sh --client-exe "${CLIENT_EXE}" --istio-dir "${ISTIO_DIR}" --istio-namespace "${ISTIO_NAMESPACE}" --namespace "${BOOKINFO_NAMESPACE}" --minikube-profile "${CLUSTER2_CONTEXT}" --kube-context "${CLUSTER2_CONTEXT}"
```
```suggestion
  source ${SCRIPT_DIR}/../install-bookinfo-demo.sh --client-exe "${CLIENT_EXE}" --istio-dir "${ISTIO_DIR}" --istio-namespace "${ISTIO_NAMESPACE}" --namespace "${BOOKINFO_NAMESPACE}" --kube-context "${CLUSTER2_CONTEXT} -tg --mongo"
```

jshaughn reviewed (approved):
See test results in prior comments. Looking good. Mazz explained my env issue with the temp file, so approved!

jshaughn commented:
@nrfox Since we are holding this for v2.15, you may want to apply my suggested code change, and possibly add my other suggestion for run-kiali.sh (in the previous comment).

nrfox commented:
Looks like GH is spotty right now so CI is failing
